source/Calamari.Terraform/TerraformConvention.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Calamari.Aws.Integration;
+using Calamari.Deployment;
+using Calamari.Deployment.Conventions;
+using Calamari.Extensions;
+using Calamari.Integration.FileSystem;
+using Calamari.Integration.Substitutions;
+using Octostache;
+
+namespace Calamari.Terraform
+{
+    public abstract class TerraformConvention : IInstallConvention
+    {
+        protected readonly ICalamariFileSystem fileSystem;
+        private readonly IFileSubstituter fileSubstituter;
+        const string DefaultTerraformFileSubstitution = "**/*.tf\n**/*.tf.json\n**/*.tfvars\n**/*.tfvars.json";
+
+        public TerraformConvention(ICalamariFileSystem fileSystem,  IFileSubstituter fileSubstituter)
+        {
+            this.fileSystem = fileSystem;
+            this.fileSubstituter = fileSubstituter;
+        }
+
+        protected abstract void Execute(RunningDeployment deployment, Dictionary<string, string> environmentVariables);
+
+        public void Install(RunningDeployment deployment)
+        {
+            InstallAsync(deployment).GetAwaiter().GetResult();
+        }
+
+        async Task InstallAsync(RunningDeployment deployment)
+        {
+            var variables = deployment.Variables;
+            var additionalFileSubstitution = variables.Get(TerraformSpecialVariables.Action.Terraform.FileSubstitution);
+            var enableNoMatchWarning = variables.Get(SpecialVariables.Package.EnableNoMatchWarning);
+
+            variables.Add(SpecialVariables.Package.EnableNoMatchWarning,
+                !String.IsNullOrEmpty(enableNoMatchWarning) ? enableNoMatchWarning : (!String.IsNullOrEmpty(additionalFileSubstitution)).ToString());
+
+            var substitutionPatterns = (DefaultTerraformFileSubstitution +
+                                        (string.IsNullOrWhiteSpace(additionalFileSubstitution)
+                                            ? String.Empty
+                                            : "\n" + additionalFileSubstitution))
+                .Split(new[] {"\r", "\n"}, StringSplitOptions.RemoveEmptyEntries);
+
+            new SubstituteInFilesConvention(fileSystem, fileSubstituter,
```

Comment:
This convention within convention doesn't feel right. Should the SubstituteInFilesConvention itself be on the terraform command like its use for all other commands?

Reply:
fixed